library/std/src/sys/pal/uefi/time.rs

```diff
@@ -63,9 +142,177 @@ pub(crate) mod system_time_internal {
+            let y_adj: u32 = (t.year as u32) + YEAR_BASE - carry;
+            let month_days: u32 = (m_adj.wrapping_add(adjust) * 62719 + 769) / 2048;
+            let leap_days: u32 = y_adj / 4 - y_adj / 100 + y_adj / 400;
+
+            // Allow days to be negative to denote days before EPOCH
+            let days: i64 =
+                (y_adj * 365 + leap_days + month_days + (t.day as u32 - 1)) as i64 - 2472632;
+
+            let localtime_epoch: i64 = days * SECS_IN_DAY
+                + (t.second as i64)
+                + (t.minute as i64) * SECS_IN_MINUTE
+                + (t.hour as i64) * SECS_IN_HOUR;
+
+            let utc_epoch: i64 = if t.timezone == r_efi::efi::UNSPECIFIED_TIMEZONE {
+                localtime_epoch
+            } else {
+                localtime_epoch + (t.timezone as i64) * SECS_IN_MINUTE
+            };
+
+            Self { secs: utc_epoch, usecs: t.nanosecond }
+        }
+
+        /// This algorithm is taken from: http://howardhinnant.github.io/date_algorithms.html
+        pub(crate) const fn to_systemtime(&self, timezone: i16, daylight: u8) -> super::SystemTime {
+            let secs: i64 = if timezone == r_efi::efi::UNSPECIFIED_TIMEZONE {
+                self.secs
+            } else {
+                self.secs - (timezone as i64) * SECS_IN_MINUTE
+            };
+
+            let (days, remaining_secs): (i64, u64) = {
+                let days = secs / SECS_IN_DAY;
+                let remaining_secs = secs % SECS_IN_DAY;
+
+                if remaining_secs < 0 {
+                    (days - 1, (SECS_IN_DAY + remaining_secs) as u64)
+                } else {
+                    (days, remaining_secs as u64)
+                }
+            };
+
+            let z = days + 719468;
+            let era = z / 146097;
+            let doe = z - (era * 146097);
+            let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
+            let mut y = yoe + era * 400;
+            let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
+            let mp = (5 * doy + 2) / 153;
+            let d = doy - (153 * mp + 2) / 5 + 1;
+            let m = if mp < 10 { mp + 3 } else { mp - 9 };
+
+            if m <= 2 {
+                y += 1;
+            }
+
+            let hour = (remaining_secs / SECS_IN_HOUR as u64) as u8;
+            let minute = ((remaining_secs % SECS_IN_HOUR as u64) / SECS_IN_MINUTE as u64) as u8;
+            let second = (remaining_secs % SECS_IN_MINUTE as u64) as u8;
+
+            super::SystemTime {
+                year: y as u16,
+                month: m as u8,
+                day: d as u8,
+                hour,
+                minute,
+                second,
+                nanosecond: self.usecs,
+                timezone,
+                daylight,
+            }
+        }
+
+        pub(crate) const fn checked_add(&self, dur: Duration) -> Self {
+            let temp: u32 = self.usecs + dur.subsec_nanos();
```

Comment:
you're combining something called `usecs` with something called `nanos` that doesn't seem to be correct?

Reply:
Isn't usec the name used for nanosecond in unix timestamp: https://docs.rs/libc/latest/libc/struct.timeval.html

Reply:
I have renamed it to `nanos` now. The reason for using `nanos` instead of normal microseconds (which is uses in unix) is because nanoseconds are stored in the UEFI Time normally.